### PR TITLE
New rspace/join removal shouldnt fail

### PR DESCRIPTION
**Pull request by lukasz-golebiewski**
## Overview
Aligns join removal with the previous implementation - failure to remove a join shouldn't raise errors

### JIRA ticket:
https://rchain.atlassian.net/browse/RCHAIN-3256

### Notes
<sup>_Optional. Add any notes on caveats, approaches you tried that didn't work, or anything else._</sup>

### Please make sure that this PR:
- [x] is at most 200 lines of code (excluding tests),
- [x] meets [RChain development coding standards](https://rchain.atlassian.net/wiki/spaces/DOC/pages/28082177/Coding+Standards),
- [x] includes tests for all added features,
- [x] has a reviewer assigned,
- [x] has [all commits signed](https://rchain.atlassian.net/wiki/spaces/DOC/pages/498630673/How+to+sign+commits+to+rchain+rchain).

### [Bors](https://bors.tech/) cheat-sheet:

- `bors r+` runs integration tests and merges the PR (if it's approved),
- `bors try` runs integration tests for the PR,
- `bors delegate+` enables non-maintainer PR authors to run the above.

### CODE REVIEW

File: rspace/src/test/scala/coop/rchain/rspace/HotStoreSpec.scala [hunk 1]
```diff
@@ -660,7 +660,7 @@ trait HotStoreSpec[F[_], M[_]] extends FlatSpec with Matchers with GeneratorDriv
                   )
               res   <- hotStore.removeJoin(channel, toRemove).attempt
               cache <- state.read
-              _     <- checkRemoval(res, cache.joins(channel), cachedJoins, index)
+              _     <- checkRemoval(res, cache.joins(channel), cachedJoins, index, shouldFail = false)
```
Comment (odeac) [inline rspace/src/test/scala/coop/rchain/rspace/HotStoreSpec.scala]: parameters like `shouldFail` make the assertion less readable. I'd rather duplicate the assertions

File: rspace/src/test/scala/coop/rchain/rspace/HotStoreSpec.scala [hunk 2]
```diff
@@ -850,10 +851,14 @@ trait HotStoreSpec[F[_], M[_]] extends FlatSpec with Matchers with GeneratorDriv
       res: Either[Throwable, Unit],
       actual: Seq[T],
       initial: Seq[T],
-      index: Int
+      index: Int,
```
Comment (odeac) [inline rspace/src/test/scala/coop/rchain/rspace/HotStoreSpec.scala]: This function becomes even more unreadable. The name `checkRemoval` doesn't say much about what it does.

File: rspace/src/test/scala/coop/rchain/rspace/HotStoreSpec.scala (hunk 1 above)
Comment (lukasz-golebiewski) [reply rspace/src/test/scala/coop/rchain/rspace/HotStoreSpec.scala]: Sure, good idea

File: rspace/src/main/scala/coop/rchain/rspace/HotStore.scala [hunk 3]
```diff
@@ -190,13 +190,13 @@ private class InMemHotStore[F[_]: Sync, C, P, A, K](
...
             val index = cache.joins(channel).indexOf(join)
-            removeIndex(cache.joins(channel), index) >>= { updated =>
+            if (index != -1) removeIndex(cache.joins(channel), index) >>= { updated =>
               Sync[F]
                 .delay {
                   cache.joins.put(channel, updated)
                 }
                 .map(_ => cache)
-            }
+            } else cache.pure[F]
```
Comment (KentShikama) [inline rspace/src/main/scala/coop/rchain/rspace/HotStore.scala]: What does index represent?

Comment (KentShikama) [reply rspace/src/main/scala/coop/rchain/rspace/HotStore.scala]: I guess why does it matter that the index is -1 for removal?

Comment (lukasz-golebiewski) [reply rspace/src/main/scala/coop/rchain/rspace/HotStore.scala]: It's the index of the element in the sequence. If it's `-1` then the element is not present (as per contract of `indexOf`) and `removeIndex` will raise an error.

File: rspace/src/main/scala/coop/rchain/rspace/HotStore.scala [hunk 4]
```diff
@@ -190,13 +190,13 @@ private class InMemHotStore[F[_]: Sync, C, P, A, K](
       continuations <- getContinuations(join)
       _ <- if (continuations.isEmpty) S.flatModify { cache =>
             val index = cache.joins(channel).indexOf(join)
-            removeIndex(cache.joins(channel), index) >>= { updated =>
+            if (index != -1) removeIndex(cache.joins(channel), index) >>= { updated =>
               Sync[F]
                 .delay {
                   cache.joins.put(channel, updated)
                 }
                 .map(_ => cache)
```
Comment (rabbitonweb) [inline rspace/src/main/scala/coop/rchain/rspace/HotStore.scala]: consider 
```
    Sync[F] .delay(cache.joins.put(channel, updated)).as(cache)
```

**Pull request merged**